File: src/main/java/com/tambapps/fft4j/complex/vector/ImmutableCVector.java
package com.tambapps.fft4j.complex.vector;

import com.tambapps.fft4j.complex.Complex;

/**
 * An immutable CVector
 */
public final class ImmutableCVector extends ArrayCVector {

  ImmutableCVector(Complex[] values) {
    super(values);
  }

  @Override
  public void setAt(int i, Complex value) {
    throw new UnsupportedOperationException("Cannot modify immutable vector");
  }

  public static CVector of(Complex... values) {
    return new ImmutableCVector(values);
  }

}

File: src/main/kotlin/com/tambapps/fft4j/complex/vector/ImmutableCVector.kt
package com.tambapps.fft4j.complex.vector

import com.tambapps.fft4j.complex.Complex
import java.lang.UnsupportedOperationException

class ImmutableCVector(elements: Array<Complex>) : ArrayCVector(elements) {

    override fun set(i: Int, value: Complex) {
        throw UnsupportedOperationException("Cannot modify an immutable vector")
    }

    companion object {

        fun of(vararg values: Complex): ImmutableCVector {
            return ImmutableCVector(arrayOf(*values))
        }
    }
}